keps/sig-network/3070-reserved-service-ip-range/README.md

```diff
@@ -0,0 +1,456 @@
+    192.168.0.1     192.168.1.1                                 192.168.255.254
+
+Example 2:
+
+- Service IP CIDR: 192.168.0.0/22
+- Range Size: 2^10 - 2 = 1022
+- Band Offset: min(max(16,1024/16),256) = min(max(16,64),256) = 64
+- Static band start: 192.168.0.1
+- Static band ends: 192.168.0.64
+
+           ┌─────────────┬─────────────────────────────────────────────┐
+           │   static    │                    dynamic                  │
+           └─────────────┴─────────────────────────────────────────────┘
+
+           ◄────────────► ◄────────────────────────────────────────────►
+    192.168.0.1     192.168.0.64                                 192.168.3.254
+
+Example 3 Small Subnet:
+
+- Service IP CIDR: 192.168.0.0/26
+- Range Size: 2^6 - 2 = 62
+- Band Offset: min(max(16,64/16),256) = min(max(16,4),256) = 16
+- Static band start: 192.168.0.1
+- Static band ends: 192.168.0.16
+
+           ┌─────────────┬─────────────────────────────────────────────┐
+           │   static    │                    dynamic                  │
+           └─────────────┴─────────────────────────────────────────────┘
+
+           ◄────────────► ◄────────────────────────────────────────────►
+    192.168.0.1     192.168.0.16                                 192.168.0.62
+
+
+```go
+// randomScanReservedStrategy choose a random address from the provided big.Int and then scans
+// forward looking for the next available address. The big.Int range is subdivided so it will try
+// to allocate first from the reserved block for dynamic allocated addresses (it will wrap around
+// back to the start of the dynamic subrange if necessary). If there is no free address it will
+// try to allocate one from the reserved block for static allocated addresses too.
+type randomScanReservedStrategy struct {
+	rand     *rand.Rand
+	reserved int
+}
+```
+
+### Test Plan
+
+This feature doesn't modify the cluster behavior, only the order on which dynamic IP are assigned to Services,
+there is no need for e2e or integration tests, unit tests are enough.
+
+### Graduation Criteria
+
+#### Alpha
+
+- Feature implemented behind a feature flag
+
+#### Beta
+
+- Gather feedback from developers and community.
+- No issues reported.
+
+#### GA
+
+- No issues reported during two releases.
+
+### Upgrade / Downgrade Strategy
+
+The feature only changes the allocation strategy, the bitmaps and the underlay logic remain the same,
+guaranteeing full compatibility and no risk rolling back the feature.
+
+### Version Skew Strategy
+
+Same as with upgrade, there is no change in the bitmaps, keeping the feature 100% compatible.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+- [x] Feature gate (also fill in values in `kep.yaml`)
+  - Feature gate name: ServiceIPStaticSubrange
+  - Components depending on the feature gate: kube-apiserver
+- [x] Other
```

Comment:
Hmm - do we really need it?

We will have a feature gate (changing which requires restart of kube-apiserver anyway).
So can't we just initialize the proper allocation strategy based on feature-gate? [Then you don't need anything else...]

Reply:
I misunderstood this question